packages/graphql-codegen-cli/tests/watcher-test-helpers/setup-mock-watcher.ts

```diff
@@ -0,0 +1,249 @@
+     * Call this functon to stop the mock watching promise (which otherwise will not terminate).
+     *
+     * This _must be called_ at the end of each test to avoid unhandled promises.
+     *
+     * Note that the assertion helper `assertBuildTriggers` will call `await stopWatching()`,
+     * so any test using that helper does not need to call it.
+     */
+    stopWatching,
+    /**
+     * Promise that is pending as long as the watcher is running.
+     *
+     * There should be no need to manually await this, because `await stopWatching()`
+     * will also wait for this same `runningWatcher` promise to resolve.
+     */
+    runningWatcher,
+    /**
+     * Asynchronous function for dispatching file change events,
+     * _which only resolves after the {@link ParcelWatcher.SubscribeCallback | subscription callback}
+     * has completed consuming the event._
+     */
+    dispatchChange,
+    /**
+     * Spy on the value of the asynchronous {@link ParcelWatcher.SubscribeCallback}
+     * that the implementing code provided as an argument
+     * to {@link ParcelWatcher.subscribe | `@parcel/watch.subscribe`}
+     *
+     * This function is called for _every_ change event, so it's a useful spy
+     * for making assertions about events that did _not_ call {@link onWatchTriggered}
+     *
+     * NOTE: Our mock does not implement {@link ParcelWatcher.Options}`["ignore"]`
+     * logic, so even if a path would be ignored, this spy will still be called.
+     * But it's possible to make assertions on ignored paths/globs by testing
+     * their values in {@link subscribeOpts} (or using the `assertBuildTriggers` helper).
+     */
+    subscribeCallbackSpy,
+    /**
+     * Mocked function implementing {@link ParcelWatcher.AsyncSubscription}`['unsubscribe']`
+     * to spy on `subscription.unsubscribe()` which should be called by implementing
+     * code when closing the subscription
+     */
+    unsubscribeSpy,
+    /**
+     * The argument that was provided to {@link ParcelWatcher.subscribe | @parcel/watch.subscribe}
+     * which indicates the directory for the Parcel Watcher to watch.
+     */
+    watchDirectory,
+    /**
+     * The {@link ParcelWatcher.Options} argument that may have been provided to
+     * {@link ParcelWatcher.subscribe | @parcel/watch.subscribe}, which will
+     * include, for example, the `ignore` key that contains ignored paths and globs that
+     * should never cause the subscription callback to be called.
+     *
+     * NOTE: This mock does _not_ implement the Parcel Watcher `shouldIgnore` check,
+     * because that's implemented by Parcel Watcher in C++ and there is no sense
+     * duplicating it in JS. So if you want to make assertions about ignored paths,
+     * you should limit it to assertions about which paths end up in `subscribeOpts.ignore`,
+     * and otherwise assume that Parcel Watcher will work as expected. For making
+     * these assertions, see the `assertBuildTriggers` helper.
+     */
```

Comment:
Just highlighting this point about the mock: we don't implement ParcelWatcher's `shouldIgnore`, so even if paths/globs are passed to the `ignore` pattern, the `subscribe` callback will still be executed in test mode. To make up for this, we have assertion helpers for making assertions about which paths/globs are passed to `ignore`, and we just assume `ParcelWatcher` will implement its behavior correctly.